### PR TITLE
DEL-3317: Add the runtime descriptor generation mechanism on the distro assembly

### PR DESCRIPTION
**Pull request by garodriguezlp**
- Building a new suit for the project (project overhauling)

### CODE REVIEW

File: src/it/verify-mojo-failing-it-test-project/src/assembly/distribution.xml [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+    <formats>
+        <format>zip</format>
+    </formats>
+    <files>
+        <file>
+            <source>src/main/resources/assembly-content/README.md</source>
+        </file>
+        <file>
+            <source>src/main/resources/assembly-content/foo.conf</source>
```
Comment (aiannucci) [inline src/it/verify-mojo-failing-it-test-project/src/assembly/distribution.xml:15]: foo is a forbidden term

File: pom.xml [hunk 2]
```diff
@@ -1,85 +1,169 @@
...
     <artifactId>mule-assembly-verifier</artifactId>
-    <version>1.5-SNAPSHOT</version>
+    <version>2.0.0-SNAPSHOT</version>
     <packaging>maven-plugin</packaging>
     <name>Mule Assembly Verifier Plugin</name>
 
+    <!-- //@todo[@theales]: What do you think about this new description? -->
     <description>
-        Verify contents of the output assembly against a controlled set of libraries.
+        Provide goals for performing important tasks in the Mule's distro assembly lifecycle like verify its contents or
```
Comment (aiannucci) [inline pom.xml:15]: "Provide goals" is redundant, since every maven plugin provides goals

Comment (aiannucci) [inline pom.xml:15]:
```suggestion
        Provide goals for performing important tasks in the Mule's distribution assembly lifecycle like verifying its contents or
```

File: README.md [hunk 3]
```diff
@@ -1,3 +1,4 @@
+[comment]: <> (//@todo: Should we add something about the new Mojo?)
```
Comment (aiannucci) [inline README.md:1]: Of course

**Review by aiannucci — CHANGES_REQUESTED** [review]

File: README.md [hunk 4]
```diff
@@ -166,7 +167,7 @@ Available options:
 ----------------------------------------------------------------------------------------------------
 |**Name**               |**Type**               |**Default**                        |**Description**|
 |-----------------------|-----------------------|-----------------------------------|-------------|
-|whitelist              |File                   |assembly-whitelist.txt             |Validation template location|
+|allowlist              |File                   |assembly-allowlist.txt             |Validation template location|
```
[the remark this answers is not in the document]
Comment (anosenzo) [reply README.md:170]: What do you mean that is not allowing/denying? if the artifacts don't match it makes the build to fail

File: pom.xml [hunk 5]
```diff
@@ -1,85 +1,169 @@
...
+        generating its descriptor
     </description>
 
     <properties>
-        <gmavenVersion>1.3</gmavenVersion>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <maven.compiler.source>1.8</maven.compiler.source>
+        <maven.compiler.target>1.8</maven.compiler.target>
+        <maven.version>3.3.9</maven.version>
+        <groovy.version>3.0.7</groovy.version>
```
Comment (anosenzo) [inline pom.xml:24]: Lets use latest groovy version 3.0.8

File: src/main/groovy/org/mule/tools/AssemblyContentsVerifierMojo.groovy [hunk 6]
```diff
@@ -1,62 +1,63 @@
...
+import org.apache.maven.plugins.annotations.Parameter
 import org.apache.maven.project.MavenProject
-import org.codehaus.gmaven.mojo.GroovyMojo
+
 import java.util.regex.Pattern
 import java.util.zip.ZipFile
 
 /**
- * @goal verify
+ * Mojo for verifying contents of the output assembly against a controlled set of libraries.
```
Comment (anosenzo) [inline src/main/groovy/org/mule/tools/AssemblyContentsVerifierMojo.groovy:14]: With "set of libraries" you mean the files inside the allowlist? in that case lest just say "set of files" or "against a controlled allowlist"

File: pom.xml [hunk 7]
```diff
@@ -1,85 +1,169 @@
...
+        <zip4j.version>2.9.0</zip4j.version>
+        <maven-plugin-annotations.version>3.6.0</maven-plugin-annotations.version>
+        <junit.version>4.12</junit.version>
```
[the remark this answers is not in the document]
Comment (garodriguezlp) [reply pom.xml:27]: Given that the [`maven-plugin-testing-harness`]() has not been upgraded for supporting Junit5, and that we're using a [Junit4 Rule](https://github.com/mulesoft/mule-assembly-verifier-plugin/blob/2ab0622efe49b940fe8b4ca8c0a06a4ee4b33715/src/test/groovy/org/mule/tools/AssemblyContentsVerifierMojoTest.groovy#L15) (which are not natively supported in [Junit5](https://junit.org/junit5/docs/current/user-guide/#migrating-from-junit4-rule-support)) I didn't want to give it a chance. 


Having that said, should we invest some effort in upgrading it to 5? 😅

File: pom.xml [hunk 8]
```diff
@@ -1,85 +1,169 @@
...
+        Provide goals for performing important tasks in the Mule's distro assembly lifecycle like verify its contents or
+        generating its descriptor
     </description>
 
     <properties>
-        <gmavenVersion>1.3</gmavenVersion>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <maven.compiler.source>1.8</maven.compiler.source>
+        <maven.compiler.target>1.8</maven.compiler.target>
+        <maven.version>3.3.9</maven.version>
```
[the remark this answers is not in the document]
Comment (garodriguezlp) [reply pom.xml]: No. But, it's already upgraded ;)

File: src/it/verify-mojo-failing-it-test-project/assembly-allowlist.txt [hunk 9]
```diff
@@ -0,0 +1,18 @@
+#
+# root
+#
+/foo-assembly-${productVersion}
```
[the remark this answers is not in the document]
Comment (garodriguezlp) [reply src/it/verify-mojo-failing-it-test-project/assembly-allowlist.txt:4]: Nooooo... Why...  I didn't know that before. 😭  

Good catch! Great, Thanks Ale!! I'll make the adjustment

[the remark this answers is not in the document]
Comment (garodriguezlp) [reply src/it/verify-mojo-failing-it-test-project/assembly-allowlist.txt]: Done!

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (garodriguezlp) [reply README.md:170]: I think `allowlist` works fine for this case given that:

- It's the suggested replacement of the `whitelist` according to the [ES Terminology recommendations](https://engineering-services.kbuild.msap.io/tools/github/TERMINOLOGY_CHECK.html#terms-violations)
- Somehow we're defining in that list the files that are allowed to be inside the assembly

File: src/it/verify-mojo-failing-it-test-project/src/assembly/distribution.xml (hunk 1 above)
Comment (garodriguezlp) [reply src/it/verify-mojo-failing-it-test-project/src/assembly/distribution.xml]: Done!

File: pom.xml (hunk 2 above)
Comment (garodriguezlp) [reply pom.xml]: Inspiration was taken from [here](https://mvnrepository.com/artifact/org.apache.maven.plugins) 👇  😅   😜

image

Does it look better now?

File: README.md (hunk 3 above)
Comment (garodriguezlp) [reply README.md:1]: Ale, are we going to merge this PR first and then create another for the new Mojo?

If so, I think that addition to the README should be part of the PR where the new Mojo gets introduced

File: src/main/groovy/org/mule/tools/AssemblyContentsVerifierMojo.groovy (hunk 6 above)
Comment (garodriguezlp) [reply src/main/groovy/org/mule/tools/AssemblyContentsVerifierMojo.groovy]: Better now? 😅

**Review by aiannucci — DISMISSED** [review]

**Review by anosenzo — APPROVED** [review]

**Pull request merged**